packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,186 @@
+import { UseFormWatch, UseFormSetValue, UseFormGetValues } from 'react-hook-form';
+import { TASK_PRODUCT_FIELD_NAMES, TaskProductFormFields, UnitOption } from '../types';
+import { getUnitOptionMap } from '../../../../util/convert-units/getUnitOptionMap';
+import { convertFn, getDefaultUnit, location_area } from '../../../../util/convert-units/unit';
+import { roundToTwoDecimal } from '../../../../util';
+import { convert } from '../../../../util/convert-units/convert';
+
+interface UseQuantityApplicationRate {
+  total_area: number;
+  total_area_unit: 'm2' | 'ha' | 'ft2' | 'ac'; // as defined in location_area
+  system: 'metric' | 'imperial';
+  watch: UseFormWatch<TaskProductFormFields>;
+  setValue: UseFormSetValue<TaskProductFormFields>;
+  getValues: UseFormGetValues<TaskProductFormFields>;
+}
+
+export const useQuantityApplicationRate = ({
+  total_area,
+  total_area_unit,
+  system,
+  watch,
+  setValue,
+  getValues,
+}: UseQuantityApplicationRate) => {
+  const application_area = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA);
+  const application_area_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA_UNIT);
+  const weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.WEIGHT_UNIT);
+  const volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.VOLUME_UNIT);
+  const application_rate_weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT_UNIT);
+  const application_rate_volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME_UNIT);
+
+  const [isWeight, setIsWeight] = useState(() => {
+    const volumeValue = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME);
+    return isNaN(Number(volumeValue));
+  });
+
+  const toggleMeasure = () => {
+    setIsWeight((prev) => !prev);
+  };
+
+  /* Set initial application area to area total area */
+  useEffect(() => {
+    if (!application_area) {
+      setValue(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA, total_area);
+
+      /* display user-selected unit initially */
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA_UNIT,
+        getUnitOptionMap()[total_area_unit] as UnitOption,
+      );
+    }
+
+    /* update unit of application area and string */
+    onPercentLocationChange();
+  }, []);
+
+  /* Update application area + rate based on percent of location */
+  const onPercentLocationChange = () => {
+    const percent_of_location = getValues(TASK_PRODUCT_FIELD_NAMES.PERCENT_OF_LOCATION);
+
+    const calculatedArea = (total_area * (percent_of_location || 100)) / 100;
+    setValue(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA, calculatedArea);
+
+    /* set unit of the total area component according to default breakpoints */
+    setValue(
+      TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA_UNIT,
+      /* @ts-ignore */
+      getUnitOptionMap()[getDefaultUnit(location_area, calculatedArea, system).displayUnit],
+      { shouldValidate: true },
+    );
+
+    /* update application rate */
+    onQuantityChange();
+  };
+
+  const onQuantityChange = () => {
```

Comment:
Looking at the function name, I expect the function to be called when the quantity is updated... (the same for `onPercentLocationChange`) `setQuantity` or `updateQuantity` might be more suitable.

Reply:
That is when it's called!

This is the function that is called when the quantity is updated, and it sets either the `application_rate_weight` or the `application_rate_volume` depending on `isWeight`. Ditto for the `onPercentLocationChange` (it's passed to the `onChange` handler the for the percent location).

If the onChange naming convention is causing confusion I can write it the other way (this would be called `updateApplicationRate` in that case). However I'm not sure I would want to rename `onPercentLocationChange`, because it would become `updateAreaAndQuantity` which seems slightly less clear to me. Do you prefer that though?